libraries/botbuilder-azure/src/cosmosDbPartitionedStorage.ts

```diff
@@ -0,0 +1,255 @@
+                // When an item is not found a CosmosException is thrown, but we want to
+                // return an empty collection so in this instance we catch and do not rethrow.
+                // Throw for any other exception.
+                if (err.code === 404) { }
+                // Throw unique error for 400s
+                else if (err.code === 400) {
+                    this.throwInformativeError(`Error initializing container. You might be using partitions in a non-partitioned DB or
+                    are not using partitions in a partitioned db that already contains partitioned data`, err);
+                } else {
+                    this.throwInformativeError('Error reading from container', err);
+                }
+            }
+        }));
+
+        return storeItems;
+    }
+
+    public async write(changes: StoreItems): Promise<void> {
+        if (!changes) { throw new ReferenceError(`Changes are required when writing.`); }
+        else if (changes.length === 0) { return; }
+
+        await this.initialize();
+
+        await Promise.all(Object.keys(changes).map(async (k: string): Promise<void> => {
+            const changesCopy: any = {...changes[k]};
+
+            // Remove eTag from JSON object that was copied from IStoreItem.
+            // The ETag information is updated as an _etag attribute in the document metadata.
+            delete changesCopy.eTag;
+            const documentChange = new DocumentStoreItem({
+                id: CosmosDbKeyEscape.escapeKey(k),
+                realId: k,
+                document: changesCopy
+            });
+
+            const eTag: string = changes[k].eTag;
+            if (!eTag || eTag === '*') {
+                // If new item or *, then insert or replace unconditionally
+                try {
+                    await this.container.items
+                        .upsert(documentChange, { disableAutomaticIdGeneration: true,  });
+                } catch (err) {
+                    this.throwInformativeError('Error upserting document', err);
+                }
+            } else if (eTag.length > 0) {
+                // If we have an etag, do opt. concurrency replace
+                try {
+                    await this.container
+                        .item(CosmosDbKeyEscape.escapeKey(k), documentChange.partitionKey)
+                        .replace(documentChange, { accessCondition: { type: 'IfMatch', condition: eTag } });
+                } catch (err) {
+                    this.throwInformativeError('Error replacing document', err);
+                }
+            } else {
+                throw new Error(`etag empty`);
+            }
+        }));
+    }
+
+    public async delete(keys: string[]): Promise<void> {
+
+        await this.initialize();
+
+        await Promise.all(keys.map(async (k: string): Promise<void> => {
+            try {
+                await this.container
+                    .item(CosmosDbKeyEscape.escapeKey(k), k)
```

Comment:
should the pk be `escapeKey(k)` here as well?

Reply:
Yep. Fixed.